### PR TITLE
[SYCL][NVPTX] Emit reqd_work_group_size attributes as NVVM annotations

### PR DESCRIPTION
**Pull request by frasercrmck**
Only emit the provided values as annotations in the LLVM IR. The NVPTX backend will pad missing values with 1s. This suits the fact that the attribute must provide as many values as the dimensionality of the work-group, and we can assume that the work-group size of unused dimensions is 1.

### CODE REVIEW

**Comment by frasercrmck** [comment]
> The code looks ok to me but this attribute is handled differently for different backends now and I do not know if that is ok (#13600). @premanandrao @steffenlarsen please weigh in here.

It's not ideal, no. I do think NVPTX.cpp is a good place to add NVPTX-specific metadata, and CodeGenFunction.cpp is a good place to add target-independent metadata.

We do need this NVPTX-specific lowering as the NVVM annotations are the only way the NVPTX backend is going to make use of the attributes - it's just ignored otherwise. I think the most regretful outcome if that we have the same information repeated in multiple different ways: see also how we are _already_ handling `[[intel::max_work_group_size]]` attribute this way. For example, `intel::max_work_group_size(4, 8, 16)` gives us:

``` llvm
define  void @foo() max_work_group_size !28 {
   ...
}

!nvvm.annotations = !{19, !20, !21}

!19 = !{ptr @_ZTSZZ4mainENKUlRN4sycl3_V17handlerEE_clES2_E1C, !"maxntidx", i32 16}
!20 = !{ptr @_ZTSZZ4mainENKUlRN4sycl3_V17handlerEE_clES2_E1C, !"maxntidy", i32 8}
!21 = !{ptr @_ZTSZZ4mainENKUlRN4sycl3_V17handlerEE_clES2_E1C, !"maxntidz", i32 4}

!28 = !{i32 16, i32 8, i32 4}
```

Perhaps one way of making it nicer would be to lower from Attribute -> Function Metadata in `CodeGenFunction.cpp`, then lower the Function Metadata to NVVM annotations in `NVPTX.cpp`. That may give a clearer lowering path, separating out responsibilities.

**Comment by frasercrmck** [comment]
Sorry, accidentally deleted the branch

File: clang/lib/CodeGen/Targets/NVPTX.cpp [hunk 1]
```diff
@@ -303,6 +303,37 @@ void NVPTXTargetCodeGenInfo::setTargetAttributes(
...
+        Ops.push_back(Dim2->getExtValue());
+      std::reverse(Ops.begin(), Ops.end());
+      Ops.append(3 - Ops.size(), std::nullopt);
+
+      // Work-group sizes (in NVVM annotations) must be positive and less than
+      // INT32_MAX, whereas SYCL can allow for larger work-group sizes (see
+      // -fno-sycl-id-queries-fit-in-int). If any dimension is too large for
+      // NVPTX, don't emit any annotation at all.
+      if (llvm::all_of(Ops, [](std::optional<int64_t> V) {
+            return !V || llvm::isUInt<31>(*V);
```
[the remark this answers is not in the document]
Comment (frasercrmck) [reply clang/lib/CodeGen/Targets/NVPTX.cpp:326]: Ah yeah, for that exact reason: we need a _positive_ 32-bit signed integer and I thought I'd compress that check into one. The PTX assembler crashes with a floating-point exception when given `UINT32_MAX`, so I had to add that filter.

**Comment by premanandrao** [comment]
> > The code looks ok to me but this attribute is handled differently for different backends now and I do not know if that is ok (#13600). @premanandrao @steffenlarsen please weigh in here.
> 
> It's not ideal, no. I do think NVPTX.cpp is a good place to add NVPTX-specific metadata, and CodeGenFunction.cpp is a good place to add target-independent metadata.
> 
> We do need this NVPTX-specific lowering as the NVVM annotations are the only way the NVPTX backend is going to make use of the attributes - it's just ignored otherwise. I think the most regretful outcome if that we have the same information repeated in multiple different ways: see also how we are _already_ handling `[[intel::max_work_group_size]]` attribute this way. For example, `intel::max_work_group_size(4, 8, 16)` gives us:
> 
> ```llvm
> define  void @foo() max_work_group_size !28 {
>    ...
> }
> 
> !nvvm.annotations = !{19, !20, !21}
> 
> !19 = !{ptr @_ZTSZZ4mainENKUlRN4sycl3_V17handlerEE_clES2_E1C, !"maxntidx", i32 16}
> !20 = !{ptr @_ZTSZZ4mainENKUlRN4sycl3_V17handlerEE_clES2_E1C, !"maxntidy", i32 8}
> !21 = !{ptr @_ZTSZZ4mainENKUlRN4sycl3_V17handlerEE_clES2_E1C, !"maxntidz", i32 4}
> 
> !28 = !{i32 16, i32 8, i32 4}
> ```
> 
> Perhaps one way of making it nicer would be to lower from Attribute -> Function Metadata in `CodeGenFunction.cpp`, then lower the Function Metadata to NVVM annotations in `NVPTX.cpp`. That may give a clearer lowering path, separating out responsibilities.

Thanks for the explanation.   I don't have a strong opinion on changing what you have, so I will approve this.   But @elizabethandrews might, but she is out until at least Monday.

**Review by premanandrao — APPROVED** [review]

**Comment by frasercrmck** [comment]
> Great! I attempted to do this a long time ago (#3755), but there were problems due to us padding the dimensions of the attribute at the time. Lovely to see it finally happening. 😄
> 
> I am not too concerned about it taking a new path for NVPTX and generating special NVVM metadata. From what I remember, the current behavior for the attribute is heavily based on the SPIR-V translator consuming the metadata that the OpenCL-related attribute set a precedence for.
> 
> An alternative could be to have a late-stage LLVM transformation pass for consuming the "reqd_work_group_size" metadata and generating these NVVM attributes. A benefit of this is that we could use this for handling "sycl-work-group-size" produced by the SYCL compile-time property alternative for this attribute.

No problem! I wasn't aware you'd already tried it, sorry.

Yeah I agree the current lowering for NVPTX (and I'm guessing AMDGPU) leaves something to be desired. I think I'll look into this in a subsequent patch, if that's alright?

**Comment by steffenlarsen** [comment]
> No problem! I wasn't aware you'd already tried it, sorry.

Don't be sorry! It was just to say that I think this is the right direction and it's great to see it actually happening. 😄 

> Yeah I agree the current lowering for NVPTX (and I'm guessing AMDGPU) leaves something to be desired. I think I'll look into this in a subsequent patch, if that's alright?

You could, but I figure it will involve moving the transformation done here to a later stage to handle both the property and the attribute the same. I don't see a problem in that though, but I'll let @elizabethandrews and @premanandrao have the final say.

**Review by smanna12 — APPROVED** [review]
LGTM. Thanks

**Comment by frasercrmck** [comment]
This PR looks ready to merge, @intel/llvm-gatekeepers - thanks!

**Pull request merged**